rclpy/rclpy/impl/rcutils_logger.py

```diff
@@ -277,6 +277,10 @@ def log(self, message, severity, **kwargs):
         from rclpy.logging import LoggingSeverity
         severity = LoggingSeverity(severity)
 
+        # Gather context info and check filters only if the severity is appropriate.
+        if not self.is_enabled_for(severity):
+            return False
+
```

Comment:
This is an obvious improvement, but I wonder if we can go a little further.  In particular, it looks like `is_enabled_for` already wraps the incoming `severity` in a LoggingSeverity object, so if we put this block of code as the very first thing in the method, we will save a tiny bit more amount of time.

Reply:
If this PR is optimized enough for the contributor's use case, and we were fine with the performance before, then I would suggest merging as is.

Reply:
I did some simple performance tests. Bear in mind, I'm not a python programmer, so there is a high chance I committed some blasphemy ;).

I wrote very simple test script:
```
import rclpy
import timeit
rclpy.init()

print(timeit.timeit('z.is_enabled_for(rclpy.logging.LoggingSeverity.FATAL)',setup='import rclpy\nz=rclpy.create_node("a").get_logger()',number=10000))
print(timeit.timeit('z.debug("ss")',setup='import rclpy\nz=rclpy.create_node("a").get_logger()',number=10000))
print(timeit.timeit('z.fatal("ss")',setup='import rclpy\nz=rclpy.create_node("a").get_logger()',number=10000))

```

For orignal code I get:
```
0.013552845921367407
9.574878340121359
9.609605725854635
```

For current modification I get:
```
0.014010991901159286
0.031755123753100634
9.542557789944112
```

If I move check above LoggingSeverity import (so import and conversion are done once if severity is not high enough) I get:
```
0.013952112291008234
0.0214669331908226
9.54146963218227
```
So there is some additional performance boost.

However, after that, I've checked what is the cost of doing this import every single time - I modified code, so LoggingSeverity class is imported in `__init__` method and stored in `self.LoggingSeverity` variable, and then it is used in is_enabled_for method. And the result is:
```
0.009306364692747593
0.017703140154480934
9.61737445788458
```

So there is some additional boost.

I'm not sure why we cannot import it on global level (I've tried but there is a message about rclpy.logging not fully initialized and possible circural dependency - this is probably the reason it is done in that way in current version), but my proposition is to:
- move severity check above import line in the log method,
- import LoggingSeverity only once in `__init__` method and store it in `self.LoggingSeverity` field, and use it in all other places in `RcutilsLogger` class

What do you think?

Reply:
I've moved the check to the begging of the function.

As for referencing LoggingSeverity, after looking into logging.py I think more elegent solution would be to break this circural dependency, either by injecting LoggingSeverity into RcutilsLogger from logging.py, or even better, by elevating LoggingSeverity enum to rcutils_logger.py or to separate module as a dependency for both logging.py and rcutils_logger.py

Reply:
Yeah, I agree that it seems much nicer to break the circular dependency and just import the module once at the top.  But I defer to @sloretz 's judgement here.

Reply:
(also, thanks for spending the time to do the benchmarks, it really helps in informing what we should do here)